lib/utils/qemu.ts

```diff
@@ -84,47 +86,50 @@ export const getQemuPath = function (arch: string) {
 	);
 };
 
-export function installQemu(arch: string) {
-	const request = require('request') as typeof import('request');
-	const fs = require('fs') as typeof import('fs');
-	const zlib = require('zlib') as typeof import('zlib');
-	const tar = require('tar-stream') as typeof import('tar-stream');
-
-	return getQemuPath(arch).then(
-		(qemuPath) =>
-			new Promise(function (resolve, reject) {
-				const installStream = fs.createWriteStream(qemuPath);
-
-				const qemuArch = balenaArchToQemuArch(arch);
-				const fileVersion = QEMU_VERSION.replace(/^v/, '').replace('+', '.');
-				const urlFile = encodeURIComponent(
-					`qemu-${fileVersion}-${qemuArch}.tar.gz`,
-				);
-				const urlVersion = encodeURIComponent(QEMU_VERSION);
-				const qemuUrl = `https://github.com/balena-io/qemu/releases/download/${urlVersion}/${urlFile}`;
-
-				const extract = tar.extract();
-				extract.on('entry', function (header, stream, next) {
-					stream.on('end', next);
-					if (header.name.includes(`qemu-${qemuArch}-static`)) {
-						stream.pipe(installStream);
-					} else {
-						stream.resume();
-					}
-				});
+async function installQemu(arch: string, qemuPath: string) {
+	const qemuArch = balenaArchToQemuArch(arch);
+	const fileVersion = QEMU_VERSION.replace(/^v/, '').replace('+', '.');
+	const urlFile = encodeURIComponent(`qemu-${fileVersion}-${qemuArch}.tar.gz`);
+	const urlVersion = encodeURIComponent(QEMU_VERSION);
+	const qemuUrl = `https://github.com/balena-io/qemu/releases/download/${urlVersion}/${urlFile}`;
 
-				return request(qemuUrl)
-					.on('error', reject)
-					.pipe(zlib.createGunzip())
-					.on('error', reject)
-					.pipe(extract)
-					.on('error', reject)
-					.on('finish', function () {
-						fs.chmodSync(qemuPath, '755');
-						resolve();
-					});
-			}),
-	);
+	const request = await import('request');
+	const fs = await import('fs');
+	const zlib = await import('zlib');
+	const tar = await import('tar-stream');
+
+	const extract = tar.extract();
+	const installStream = fs.createWriteStream(qemuPath);
+	try {
+		extract.on('entry', function (header, stream, next) {
+			stream.on('end', next);
+			if (header.name.includes(`qemu-${qemuArch}-static`)) {
+				stream.pipe(installStream);
+			} else {
+				stream.resume();
+			}
+		});
```

Comment:
Does having this part under `try` have any effect?  What could throw?

Reply:
If unnecessary, It is also certainly harmless. I wasn't sure about one specific situation: suppose the execution flow "stopped" at the `await` clause below (`await new Promise(...)`). While "stopped" there, suppose `header.name.includes()` above threw because name is undefined (assume it is possible, though it may not be). In this case, would this `try` block still catch the error? Considering that the error is thrown by a callback of `extract.on`. My gut feeling is that the error would not be caught, but I wasn't sure.

In summary, I think you are right that the `extract.on` block could be moved outside the `try` block, but also harmless to be there. If other changes are needed to the PR I can move this block outside the try-catch, but if this was the only change, I'll just let it be. :-)

Reply:
> While "stopped" there, suppose header.name.includes() above threw because name is undefined (assume it is possible, though it may not be). In this case, would this try block still catch the error? Considering that the error is thrown by a callback of extract.on. My gut feeling is that the error would not be caught, but I wasn't sure.

Yeah I think that's outside of the context of the try at that point (since the throw is happening in a callback).  I agree it's harmless, but wanted to check that there wasn't a bad assumption.

Reply:
The commit that this comment referred to was moved to PR #2207.